pvlib/pvsystem.py

```diff
@@ -1164,16 +1165,16 @@ def get_iam(self, aoi, iam_model='physical'):
             if `iam_model` is not a valid model name.
         """
         model = iam_model.lower()
-        if model in ['ashrae', 'physical', 'martin_ruiz']:
-            param_names = iam._IAM_MODEL_PARAMS[model]
-            kwargs = _build_kwargs(param_names, self.module_parameters)
-            func = getattr(iam, model)
+        if model in ['ashrae', 'physical', 'martin_ruiz', 'interp']:
+            func = getattr(iam, model)  # get function at pvlib.iam
+            # get all parameters from function signature to retrieve them from
+            # module_parameters if present
+            params = set(inspect.signature(func).parameters.keys())
+            params.discard('aoi')  # exclude aoi so it can't be repeated
+            kwargs = _build_kwargs(params, self.module_parameters)
             return func(aoi, **kwargs)
         elif model == 'sapm':
             return iam.sapm(aoi, self.module_parameters)
```

Comment:
I notice also that we don't have a way to pass `upper` through to `pvlib.iam.sapm`.  Separate issue.

Reply:
Hmmmm. I don't like many things about the current interface of these IAM functions; IMO, a good approach is shown at `physical`:
``` python
def physical(aoi, n=1.526, K=4.0, L=0.002, *, n_ar=None):
```

Note the asterisk (it means following args are [keyword-only args](https://stackoverflow.com/questions/22248396/keyword-only-parameter) just in case).

Each argument, of different [kinds](https://docs.python.org/3/library/inspect.html#inspect.Parameter.kind), parse to a real-world meaning (again, IMO).

`POSITIONAL_ONLY` (before `/`) & `POSITIONAL_OR_KEYWORD`: used for required input to run models. In the case of AOI modifiers, it is the `aoi` parameter. We see that they can also be used to specify defaults for the model params. `{aoi, n, K, L}`

`KEYWORD_ONLY`: used to specify additional inputs, that are not required in the published paper of the model. `{n_ar}`

I would still upgrade the function signature to:
``` python
def physical(aoi, /, n=1.526, K=4.0, L=0.002, *, n_ar=None):
```
So it is always the first parameter the input. This is completely unnecessary in IAM models since the only required input to work is `aoi`.

### Why this proposal?
Because of the [inspect module](https://docs.python.org/3/library/inspect.html). In this PR I used it to retrieve all the parameters needed:

```python
if model in ['ashrae', 'physical', 'martin_ruiz', 'interp']:
    func = getattr(iam, model)  # get function at pvlib.iam
    # get all parameters from function signature to retrieve them from
    # module_parameters if present
    params = set(inspect.signature(func).parameters.keys())
    params.discard('aoi')  # exclude aoi so it can't be repeated
    kwargs = _build_kwargs(params, self.module_parameters)
    return func(aoi, **kwargs)
``` 

With this broad approach to the function parameter meanings, we could test for the types of each one and infer it's use with `inspect`: we could stop maintaining `iam._IAM_MODEL_PARAMETERS`, which only is used [to infer the right model](https://github.com/pvlib/pvlib-python/pull/1832#discussion_r1304410955) as stated by @cwhanse . Except for `iam.sapm`, its interface is special but could be changed easily.

I tried not to use `iam._IAM_MODEL_PARAMETERS` in the whole code, but with no luck OFC, functions interfaces had to be changed - a breaking API change.

### In short,
I propose a normalized way of defining function signatures, where:
 * `POSITIONAL_ONLY` arguments are the required ones to get numerical values that may make sense (this is open to debate [1])
 * `POSITIONAL_OR_KEYWORD` are the ones that fit a function or a physical magnitude of interest, that usually (always?) has a default values
 * `KEYWORD_ONLY` args are completely optional to run the model
So with the `inspect` module utilities all this information is easily obtained and no more inference dicts are needed.

[1] Since also required ones are those without default values and it [can also be checked with the inspect module](https://stackoverflow.com/questions/58598059/how-to-find-the-required-arguments-of-a-function-object-in-python).

In the case of IAM module:

```python
def ashrae(aoi, /, b=0.05):
def physical(aoi, /, n=1.526, K=4.0, L=0.002, *, n_ar=None):
def martin_ruiz(aoi, /, a_r=0.16):
def martin_ruiz_diffuse(surface_tilt, /, a_r=0.16, c1=0.4244, *, c2=None):
def interp(aoi, theta_ref, iam_ref, *, method='linear', normalize=True):
def sapm(aoi, /, module, *, upper=None):  # here module should be changed to B5 ... B0
...
``` 
Or without the ` /,`, as stated per [1].

I see this got really long, may I open a discussion?

Reply:
> may I open a discussion?

Absolutely!  Better to have a dedicated thread for this.  Otherwise it will get buried soon when the PR is merged :)